### DISCUSSION TITLE
Gemspec Marshal dump for quick index generation

### DISCUSSION THREAD

**Discussion by ep4sh** [ruby/rubygems#6636] [Q&A] [answered]
Dear community,
could someone guide me on such thing as Gemspec Marshaling? I've found the "dump" method:
https://github.com/rubygems/rubygems/blob/560e37f8055b723de89dfb58823b261305b565c0/bundler/lib/bundler/yaml_serializer.rb#L13
Which is used to create bytes structure of YAML gemspec. I am not sure if I fully understand what is happening on lines above, I appreciate if somebody could explain (in English) why do we need so structure.

If I am able to read this code we just iterate over hash KV and assign V to K.

More general question - why we can't just use gemspec object as is?

PS I am not so clever at Ruby,
thanks for patience =)

**Comment by simi** [comment]
I'm not sure what are you looking for exactly, but gem specification is being serialized in `Gem::Specification#to_yaml` method.

**Reply by ep4sh** [reply]
Thank you for the reply, I'm asking about the quick index generation:
https://github.com/rubygems/rubygems/blob/560e37f8055b723de89dfb58823b261305b565c0/lib/rubygems/indexer.rb#L114

as I understood we use zlib to compress the data, and this data is a preformatted gemspec

**Reply by simi** [reply]
I still don't follow. Would you mind to explain what are you trying to achieve? What is your goal?

**Reply by ep4sh** [reply]
Sure, I'm developing my own gem registry with strongly-typed language. I was confused with metadata generation.

As I understood from the above, there we need to deflate gem specification, and this specification is not just a gemspec, but gemspec in odd format (well, of course just for me - since I just don't understand why), which started with newline:
https://github.com/rubygems/rubygems/blob/560e37f8055b723de89dfb58823b261305b565c0/bundler/lib/bundler/yaml_serializer.rb#L14
 and then some sort of Ruby magic happening, which I couldn't understand =(

I appreciate it if you could point me to the `Marshal.dump()` explanation in human-friendly words, like RFC or something like that.

Thanks in advance,
P.

**Reply by ep4sh** [reply]
Hi, Josef, do you have any ideas how to parse this logic to adopt into any other PL ?

**Reply by ep4sh** [reply, marked as answer]
Nevermind, it looks like I've found the answer - https://docs.ruby-lang.org/en/2.1.0/marshal_rdoc.html
It's really strange having all that stuff, IMO

**Reply by simi** [reply]
@ep4sh ad Marshal, you can find more at https://github.com/ruby/ruby/blob/master/doc/marshal.rdoc
